File: exercism/java/yacht/src/main/java/Yacht.java
import java.util.*;
import java.util.function.Function;
import java.util.stream.Collectors;

class Yacht {

    private static final int LOSING_SCORE = 0;
    private static final int YACHT_SCORE = 50;
    private static final int STRAIGHT_SCORE = 30;

    private int[] dices;
    private final YachtCategory yachtCategory;

    Yacht(int[] dices, YachtCategory yachtCategory) {
        this.dices = Arrays.copyOf(dices, dices.length);
        this.yachtCategory = yachtCategory;
    }

    int score() {
        switch (this.yachtCategory) {
            case YACHT:
                return this.scoreForYacht();
            case ONES:
                return this.scoreForNumbers(1);
            case TWOS:
                return this.scoreForNumbers(2);
            case THREES:
                return this.scoreForNumbers(3);
            case FOURS:
                return this.scoreForNumbers(4);
            case FIVES:
                return this.scoreForNumbers(5);
            case SIXES:
                return this.scoreForNumbers(6);
            case FULL_HOUSE:
                return this.scoreForFullHouse();
            case FOUR_OF_A_KIND:
                return this.scoreForFourOfAKind();
            case LITTLE_STRAIGHT:
                return this.scoreForStraight(Arrays.asList(1,2,3,4,5));
            case BIG_STRAIGHT:
                return this.scoreForStraight(Arrays.asList(2,3,4,5,6));
            case CHOICE:
                return this.scoreForChoice();
            default:
                return -1;
        }
    }

    private int scoreForYacht() {
        return Arrays.stream(dices).distinct().count() == 1 ? YACHT_SCORE : LOSING_SCORE;
    }

    private int scoreForNumbers(int number) {
        return (int) (Arrays.stream(dices).filter(i -> i == number).count() * number);
    }

    private int scoreForFullHouse() {
        Map<Integer, Long> groupByCount = Arrays.stream(dices).boxed().collect(Collectors.groupingBy(Function.identity(), Collectors.counting()));
        if (new HashSet<>(groupByCount.values()).equals(new HashSet<>(Arrays.asList(2L, 3L)))) {
            return Arrays.stream(dices).sum();
        }
        return LOSING_SCORE;
    }

    private int scoreForFourOfAKind() {
        Map<Integer, Long> groupByCount = Arrays.stream(dices).boxed().collect(Collectors.groupingBy(Function.identity(), Collectors.counting()));
        if (new HashSet<>(groupByCount.values()).equals(new HashSet<>(Arrays.asList(1L, 4L))) || new HashSet<>(groupByCount.values()).equals(new HashSet<>(Collections.singletonList(5L)))) {
            return groupByCount.entrySet().stream().filter(entry -> entry.getValue() == 4L || entry.getValue() == 5L).map(entry -> entry.getKey() * 4).findFirst().get();
        }
        return LOSING_SCORE;
    }

    private int scoreForStraight(List<Integer> expectedResult) {
        Set<Integer> littleStraight = new HashSet<>(expectedResult);
        Set<Integer> givenRoll = Arrays.stream(dices).boxed().collect(Collectors.toSet());
        return givenRoll.equals(littleStraight) ? STRAIGHT_SCORE : LOSING_SCORE;
    }

    private int scoreForChoice() {
        return Arrays.stream(dices).sum();
    }
}

File: exercism/python/yacht/yacht.py
"""
This exercise stub and the test suite contain several enumerated constants.

Since Python 2 does not have the enum module, the idiomatic way to write
enumerated constants has traditionally been a NAME assigned to an arbitrary,
but unique value. An integer is traditionally used because it’s memory
efficient.
It is a common practice to export both constants and functions that work with
those constants (ex. the constants in the os, subprocess and re modules).

You can learn more here: https://en.wikipedia.org/wiki/Enumerated_type
"""

from collections import Counter

# Score categories.
# Change the values as you see fit.
YACHT = 0
ONES = 1
TWOS = 2
THREES = 3
FOURS = 4
FIVES = 5
SIXES = 6
FULL_HOUSE = 7
FOUR_OF_A_KIND = 8
LITTLE_STRAIGHT = 9
BIG_STRAIGHT = 10
CHOICE = 11


def score(dice, category):
    counter = Counter(dice)

    def fun(number, counter):
        return number * counter[number]

    score_by_category = {
        YACHT: 50 if counter.most_common(1)[0][1] == 5 else 0,
        ONES: fun(1, counter),
        TWOS: fun(2, counter),
        THREES: fun(3, counter),
        FOURS: fun(4, counter),
        FIVES: fun(5, counter),
        SIXES: fun(6, counter),
        FULL_HOUSE: sum(dice) if sorted(counter.values()) == [2, 3] else 0,
        FOUR_OF_A_KIND: 4 * counter.most_common(1)[0][0]
        if counter.most_common(1)[0][1] >= 4
        else 0,
        LITTLE_STRAIGHT: 30 if sorted(dice) == list(range(1, 6)) else 0,
        BIG_STRAIGHT: 30 if sorted(dice) == list(range(2, 7)) else 0,
        CHOICE: sum(dice),
    }

    return score_by_category[category]
